echo-pubsub-core/src/main/java/com/netflix/spinnaker/echo/pubsub/PubsubMessageHandler.java

```diff
@@ -112,11 +122,29 @@ private void setMessageHandled(String messageKey, String identifier, Long retent
     });
   }
 
+  private void setMessageComplete(String messageKey, String payload, Long retentionDeadlineMillis) {
+    redisClientDelegate.withCommandsClient(c -> {
+      c.psetex(messageKey, retentionDeadlineMillis, payload);
+    });
+  }
+
+  private Boolean messageComplete(String messageKey, String value) {
+    Boolean exists = redisClientDelegate.withCommandsClient(c -> {
```

Comment:
I think this is not safely concurrent between Echo instances. Reads are non-blocking, so multiple Echoes could attempt to write the 'completed' key here, and the behavior afterwards is not clear.

Reply:
Ah good catch. Fixed.